New Func backend: First iteration

byakuren-hijiri opened the pull request:
Closes #526
Closes #619 (FunC grammar & parser by @novusnota)

### Main roadmap
1. ~~Write AST/formatter, and utilities to generate and emit Func code.~~
3. Rewrite the existing codegen using the API created on the step 1.
   It should be rewritten entirely at this step, since everything is interconnected in the old backend, and there are no any unit tests checking any specific functionality of the generated code.
4. Provide differential testing between two backends achieving good code coverage for the new backend.
    The previous backend is basically untested. It requires to write a number of Tact contracts and setup code coverage to ensure that all the implemented features of the new backend work in the same way as in the old one.
5. Refactor the new backend and remove the old backend.

### Sub-roadmap of @novusnota

The goal is to create a comprehensive FunC grammar and parsed using the Ohm parsing toolkit. Should cover FunC version 0.4.4:
- [x] `#pragma ...`
  - [x] Grammar (`grammar.ohm`)
  - [x] Semantic actions (`semantics.addOperation`)
  - [x] Unit tests (positive and negative)
- [x] `#include ...`
  - [x] Grammar
  - [x] Semantic actions
  - [x] Unit tests
- [x] `global ...`
  - [x] Grammar
  - [x] Semantic actions
  - [x] Unit tests
- [x] `const ...`
  - [x] Grammar
  - [x] Semantic actions
  - [x] Unit tests
- [x] Identifiers & Types (unexpectedly tough!)
  - [x] Grammar
  - [x] Semantic actions
  - [x] Unit tests
- [x] Comments and literals
  - [x] Grammar
  - [x] Semantic actions
  - [x] Unit tests
- [x] Functions (declarations & definitions, including asm ones):
  - [x] Grammar
  - [x] Semantic actions
  - [x] Unit tests
- [x] Statements
  - [x] Grammar
  - [x] Semantic actions
  - [x] Unit tests
- [x] Expressions (just like in C++'s parser, with all its precedences matched!)
  - [x] Grammar
  - [x] Semantic actions
  - [x] Unit tests
- [x] Done?
- [x] (maybe) Add tests with .fc files from `ton-blockchain/ton` monorepo

### Current progress
* [x] FunC syntax tree (AST + comments)
* [x] FunC syntax iterators
* [ ] FunC formatter (we can skip unit tests for now, as it will be tested as a part of the Tact compiler pipeline)
  + [ ] ~~Support JSON dumps for debugging purposes~~
  + [x] Support indentation
  + [ ] Support parens
* [x] New FunC generators (based on `src/generators/writers/write*.ts`; similar code structure if possible)
  + [x] `Expression`
  + [x] `Statement`
  + [x] `Function`
  + [x] `Literal`
  + [x] `Contract`
  + [x] Interfaces
  + [x] Primitives
  + [x] `Constant`
* [x] Rewrite utilities from the current codegen (`WriteContext`)
  + [x] Type utilities (`src/generators/resolveFunc.*.ts`)
  + [x] ABI helpers
  + [x] Routers
  + [x] Receivers
  + [x] Getters
  + [x] Serializers
  + [x] Accessors
  + [x] Stdlib
  + [x] Init serializers
  + [x] Storage functions
  + [x] Static functions
  + [x] Extenstions
  + [x] ~~"used" identifiers~~
* [x] Write the dependency management used when emitting code from `WriteContext`. The existing structure based on `.context`/`.used` calls is too imperative and makes the source code unreadable. Thus, a better design should be developed.
* [ ] Write a brief description of the structure of the generated Func code and the high-level overview of the architecture of the codegen (in e.g. `src/codegen/README.md`)
* [x] Merge with `main` incorporating upstream changes to the new backend:
  + https://github.com/tact-lang/tact/pull/615
* [ ] Provide differential testing of two backends
  + [ ] ~~Write new unit tests if needed~~ Write unit tests for the codegen, since we don't have any
  + [x] Provide functions for differential testing of two backends
  + [ ] Configure the fuzzer to support both backends
* [ ] Refactor the new backend
  + [ ] Change names and code structure inherited from the original backend
  + [ ] Extracts methods where possible; as for now it is unreadable
* [ ] Remove the previous backend

### Notes
* Running the new backend: `yarn build && NEW_CODEGEN=1 PRINT_FUNC=1 ./bin/tact --config <project.config.json>`. It will print the generated source in stdout. Unset `NEW_CODEGEN` to run the same with the old backend.
* Running the differential test: ` yarn test src/codegen/codegen.spec.ts`
* All the elements in the AST have the `FuncAst` prefix. When creating a separate packge, we could remove the `Func` part; and add it only to our project. This could be done creating a separate module that includes `index.ts` with public exports like: `export { AstComment as FuncAstComment } from '@func-package-name';`

---

- [ ] I have updated CHANGELOG.md
- [ ] I have documented my contribution in Tact Docs: https://github.com/tact-lang/tact-docs/pull/PR-NUMBER
- [ ] I have added tests to demonstrate the contribution is correctly implemented: this usually includes both positive and negative tests, showing the happy path(s) and featuring intentionally broken cases
- [ ] I have run all the tests locally and no test failure was reported
- [ ] I have run the linter, formatter and spellchecker
- [ ] I did not do unrelated and/or undiscussed refactorings

byakuren-hijiri committed:
feat: Initial draft of the new backend.

+ Func syntax tree
+ Essential functionality Func formatter
+ A new `codegen` package that includes the updated backend
+ Hacks in the compilation pipeline that allow to use both backends
  which is needed for differential testing.

byakuren-hijiri committed:
feat(codegen): Support all `ops`

byakuren-hijiri committed:
feat(codegen): Support unary expressions + refactor

byakuren-hijiri committed:
feat(func): Support regular assigns

byakuren-hijiri committed:
feat(codegen): Support assign statements

byakuren-hijiri committed:
chore(codegen): Refactor statements

byakuren-hijiri committed:
feat(codegen): Support field access

byakuren-hijiri committed:
fix(syntax): Remove assignments from available binary operations

byakuren-hijiri committed:
chore(syntax): Refactor

byakuren-hijiri committed:
feat(codegen): Support all the binary expressions

byakuren-hijiri committed:
chore(function): Reuse `resolveFuncType`

byakuren-hijiri committed:
feat(codegen): Constructor functions + relevant changes in AST

byakuren-hijiri committed:
chore(codegen): Func example

byakuren-hijiri committed:
feat(codegen): Introduce `CodegenContext`

Needed for bottom-up AST generation; e.g. function definiton might be
generated from the expression that calls it.

byakuren-hijiri committed:
feat(codegen): Calls to constructor functions + context-related logic

byakuren-hijiri committed:
feat(codegen): Support struct values/objects

byakuren-hijiri committed:
fix(codegen): Minor errors

byakuren-hijiri committed:
feat(ast): Multiline comments

byakuren-hijiri committed:
fix(formatter): Function signatures

byakuren-hijiri committed:
feat(formatter): Indentation

byakuren-hijiri committed:
chore(formatter): Extra newlines

byakuren-hijiri committed:
feat(codegen): Comments, pragmas, includes in the generated contract

byakuren-hijiri committed:
fix(formatter): VarDef w/ local type inference

byakuren-hijiri committed:
fix(codegen): tuples=>tensors as return values of functions

byakuren-hijiri committed:
feat(codegen): contract=>module; draft main contract generation logic

byakuren-hijiri committed:
feat(codegen): Support `init_of` and ternary expressions

byakuren-hijiri committed:
feat(syntax): Function definition/declaration

byakuren-hijiri committed:
feat(codegen): Support multi-file generation

byakuren-hijiri committed:
feat(codegen): `lazy_deployment_completed` and `get_abi_ipfs`

byakuren-hijiri committed:
feat(syntax): Support string literal types

byakuren-hijiri committed:
feat(codegen): Generate `supported_interfaces`

byakuren-hijiri committed:
feat(syntax): Introduce `constructors`

byakuren-hijiri committed:
feat(syntax+fmt): Support hex int literals; refactor

byakuren-hijiri committed:
fix(fmt): Missing semicolons

byakuren-hijiri committed:
feat(codegen): Routers; handlers for receivers

byakuren-hijiri committed:
feat(syntax+codegen): Support new AST constructors

byakuren-hijiri committed:
feat(syntax): Support different comment styles

byakuren-hijiri committed:
chore(syntax): %s/varDef/vardef/g for consistency

byakuren-hijiri committed:
feat(syntax): Utilize `FuncAstIdExpr` for names instead of raw strings

byakuren-hijiri committed:
feat(tests): Differential tests based on jest' diff utility

byakuren-hijiri committed:
fix(fmt): Minor fixes

byakuren-hijiri committed:
fix(fmt): Trailing whitespaces in comments

byakuren-hijiri committed:
fix(codegen): Typo

byakuren-hijiri committed:
feat(formatter): line length limit + multiline tensor expressions

byakuren-hijiri committed:
feat(syntax+formatter): Support function receivers

byakuren-hijiri committed:
fix(codegen): Correct function receivers

byakuren-hijiri committed:
chore(compile): Optional debug prints for codegen output

byakuren-hijiri committed:
fix(codegen): Get rid of extra function attributes

byakuren-hijiri committed:
chore(codegen): Replace the rest of ad-hoc initializers w/ constructors

byakuren-hijiri committed:
feat(syntax): Extra check for implicit `Object`->`string` conversion

byakuren-hijiri committed:
feat(syntax): `asm` functions

byakuren-hijiri committed:
feat(codegen): Separate `Literal` generator; string literals support

byakuren-hijiri committed:
feat(codegen): Support `Address`

byakuren-hijiri committed:
feat(codegen): Support `Cell` and `Comment` literals

byakuren-hijiri committed:
chore(syntax): More informative error message

byakuren-hijiri committed:
fix(codegen): `[object Object]` in output

byakuren-hijiri committed:
fix(codegen): Comment position

byakuren-hijiri committed:
feat(syntax): Support extra CR

byakuren-hijiri committed:
feat(codegen): Support receivers

byakuren-hijiri committed:
feat(syntax): Support value in `method_id` attributes

byakuren-hijiri committed:
fix(codegen): Comment position

byakuren-hijiri committed:
feat(codegen): Support getters

byakuren-hijiri committed:
feat(codegen): CRs in routers code to look as the old backend

byakuren-hijiri committed:
feat(codegen): Context management to replicate some `Writer` logic

novusnota commented on src/codegen/context.ts:
```diff
@@ -1,11 +1,48 @@
+ * Basically, it tells in which file the context value should be located in the
+ * generated Func code.
+ *
+ * TODO: Should be refactored; `type` seems to be redundant
```
TypeScript does type erasure upon compilation, so in runtime this may be needed :)

byakuren-hijiri committed:
feat(codegen): Support internal receivers

byakuren-hijiri committed:
fix(codgen): Bug in the old backend

byakuren-hijiri committed:
feat(codegen): Support external receivers

byakuren-hijiri committed:
fix(codegen): Arguments of `throw_unless`

See: https://github.com/tact-lang/tact/pull/604

byakuren-hijiri committed:
feat(formatter): Set parens around non-trivial unary expressions

byakuren-hijiri committed:
feat(codegen): More convinient context API for adding function

byakuren-hijiri committed:
feat(codegen): Support init and initChild functions

byakuren-hijiri committed:
feat(codegen): Add primitives

byakuren-hijiri committed:
chore(codegen): Revisit the structure of `writeProgram`

byakuren-hijiri committed:
feat(codegen): Support native Func code

byakuren-hijiri committed:
feat(codegen): Generate the constants file

byakuren-hijiri committed:
feat(codegen): Support storage

byakuren-hijiri committed:
fix(formatter): Extra space in fun declarations w/o attrs

byakuren-hijiri committed:
feat(syntax): Func syntax iterator

byakuren-hijiri committed:
feat(codegen): Automatic depdendency detection

Extended the logic of the context to automatically detect dependencies
between functions based on the given AST. It replicates the previous
`WriterContext` logic, but makes things less "imperative", removing
state variables.

byakuren-hijiri committed:
chore(codegen): Skip dependencies check

We skip this, because we automatically add the standard Func functions as dependencies, which are not present in the context.

byakuren-hijiri committed:
feat(codegen): Introduce hacks in dependency management to overcome missing functions

byakuren-hijiri committed:
chore(codegen): Restructure, to make `writeAll` more explicit

byakuren-hijiri committed:
feat(syntax): Support variables unpacking

byakuren-hijiri committed:
feat(codgen): WIP stdlib translation to showcase how it looks like

novusnota committed:
test: small subset of functions in stdlib for testing FunC grammar

novusnota committed:
WIP: FunC grammar & parser

Needs statements, expressions, and a bit of polishing. Then, I'll add
semantical analysis and exhaustive tests of all the items.

novusnota committed:
feat(func-parser): mostly finished the expressions and fixed types

novusnota committed:
feat: tested top-level things, added more tests for identifiers

Fully tested:
* Asm Functions
* Globals
* Pragmas
* Constants

Partially tested: Function definitions and declarations
Next up: Statements

Broght over identifier tests from the funcId fixing PRs. Almost re-wrote
the universally correct identifier recognition, stuck with underscores
for a bit. Will move past that and finilize all other things, up to (and
perhaps including) expressions + semantics. In the background I'll try
to polish identifiers some more :)

novusnota committed:
feat(func-parser): 1/3 of types and semantics

Later on, `FuncAst...` types from `grammar.ts` need to be merged with
ones in `syntax.ts`, in favor of the former

novusnota committed:
test(func-parser): covered everything but expressions in tests

Not all pass due to the complicated things with polymorphic types and
identifiers. To be resolved tomorrow, with semantic actions and
expressions.

novusnota committed:
fix: statements, expressions and types in different contexts

novusnota committed:
test(func-parser): one more sample of an interesting asm function

novusnota committed:
fix(func-parser): multi-line and single-line strings can be placed after each other in asm functions

feat(func-parser): divide polymorphic and non-polymorphic, uniform types

novusnota committed:
feat/chore(func-parser): types, loc and formatting

novusnota committed:
fix(func-parser): adjust expressions and identifiers

novusnota committed:
func(func-parser): grammar is finally completed, and most tests pass!

However, I still have to restrain the identifiers — they parse too much.
Then, the `grammar.ts` will be swiftly completed and the parser would be
done. I'll do that ASAP

novusnota committed:
fix(func-parser): fixed most of the bugs, working on identifiers starting with underscores

...and/or similar things

novusnota committed:
refactor(func-parser): types, expressions, and whitespace padding

Next up: two highest-precedence sets of expressions and rework of
identifiers. Then, if that works out, semantic actions (finally!). If
even that won't, I'm calling FunC "context-sensitive", getting rid of
Ohm and swiftly hand-rolling a recursive descent parsed based on the C++
sources of FunC. *sigh*

novusnota committed:
feat(func-parser): grammar is finally complete and able to parse stdlib.fc and Tact's stdlib

Note, that generic types in variable declarations are temporarily
omitted, as they turn FunC into a context-sensitive grammar. This could
be resolved in the future, by either doing some hacks during syntax
analysis, or by introducing, say, a `decl` keyword into FunC to prefix
all variable declarations: `decl var/_/type(s) ids...`

novusnota committed:
feat(func-parser): worked through most of the types, got to complete the AST construction (mostly typing stuff in), and we're done!

novusnota committed:
feat(func-parser): 2/3 of the AST

novusnota committed:
feat(func-parser): completed! :rocket:

novusnota committed:
fix(func-parser): bug squashed and lots of tests added

novusnota committed:
chore(func-parser): change the note

novusnota committed:
chore(func-parser): simplify `upperId`' regex and reformat `grammar.ts`

Tiny nits

novusnota committed:
chore(func-parser): spelling

jubnzv committed:
fix(generator): Fix compilation errors

jubnzv committed:
feat(grammar): Add dummy definitions

jubnzv committed:
chore(grammar): Extract type

novusnota committed:
chore(func-parser): remove redundant snapshots and rename couple things in `grammar.ts`'

jubnzv committed:
feat(func/grammar): Separate hex literal

Needed in the FunC codegen

jubnzv committed:
fix(grammar): Set correct types for assignment' `lhs` and `rhs`

jubnzv committed:
feat(syntaxConstructors): Support new `grammar.ts`

jubnzv committed:
feat(codegen): Adapted the codegen

jubnzv committed:
feat(iterators): Support new grammar

novusnota commented on src/func/grammar.ts:
```diff
@@ -1407,23 +1418,33 @@ export type FuncAstComment = FuncAstCommentSingleLine | FuncAstCommentMultiLine;
 export type FuncAstCommentSingleLine = {
     kind: "comment_singleline";
     line: string;
+    style: ";" | ";;";
```
@jubnzv Single `;` is not a comment, but an empty statement (FunC :shrug:)! Single-line comments must start with `;;`

Otherwise — awesome changes!

jubnzv replied:
Thanks! That's right

jubnzv committed:
fix(grammar): Forbid ;-style comments

jubnzv committed:
fix(codegen): Update some functionality

jubnzv committed:
feat(func): Half-backed `pp` for the new AST

jubnzv committed:
feat: Finalize new AST translation

jubnzv committed:
fix(syntaxConstructors): Allow empty `method_id`

jubnzv committed:
fix(pp): Support `unit`

jubnzv committed:
fix(package.json): Include func grammar bundle to dist

jubnzv committed:
fix(pp): Global types

jubnzv committed:
feat(codegen): Add stdlib

jubnzv committed:
fix(stdlib): Hack for Func parser

jubnzv committed:
fix(stdlib): Missing brackets

jubnzv committed:
chore(ctx): Add missing `parse`

jubnzv committed:
chore(stdlib): simplify

jubnzv committed:
chore(pp): ppTy as a public interface

jubnzv committed:
feat(codegen): Add accessors

jubnzv committed:
fix(accessors): Syntax errors

jubnzv committed:
feat(codegen): Serializers

jubnzv committed:
fix(util): `funcInitIdOf`

jubnzv committed:
feat(codegen): Init serializers

jubnzv committed:
feat(codegen): Storage functions

jubnzv committed:
feat(codegen): Add static functions

jubnzv committed:
chore(codegen): Add missing file

jubnzv committed:
feat(codegen): Support extensions

jubnzv committed:
feat(grammar): Support `try {} catch (_) {}`

jubnzv committed:
feat(codegen): Support all the statements

jubnzv committed:
feat(pp): Refine API; more public methods

jubnzv committed:
feat(pp): Support type vars

jubnzv committed:
chore(Writer): Replace `#` annotations

Needed to test the backend leveraging structural typing in TS

jubnzv committed:
feat(func): Expose public functions

jubnzv committed:
chore(codegen): Remove tests; they'll be rewritten

jubnzv committed:
Merge remote-tracking branch 'origin/main' into 526-func-codegen

jubnzv committed:
feat: New codegen copied from the main version; `writeStdlib` is ported

jubnzv committed:
fix(codegen): Remove extra attributes

jubnzv committed:
chore(codegen): Roll-back porting `writeStdlib` because of FunC parser erorrs

jubnzv committed:
chore(syntaxConstructors): Set `if` instead of `ifnot` by default

jubnzv committed:
feat(test): Improve readability

jubnzv committed:
feat(test): Don't add extra .tact to paths in the generated config

jubnzv committed:
feat(test): Don't consider newlines in differential testing

Both backends generated some extra garbage.

jubnzv committed:
fix(pp): Don't add extra crs

jubnzv committed:
feat(codegen): Port `writeRouter`

novusnota committed:
feat: support `catch(_)` clauses

novusnota committed:
fix(tests): Hack for FunC parser

Still idk, why things went wrong with that identifier there, as
identifiers elsewhere work just fine

novusnota committed:
fix(func-parser): Correct recognition of function and method calls

And semantically accurate function arguments (just tensors), albeit at
the cost of some expressiveness of FunC

jubnzv committed:
feat(test): Allow lowercase test names

jubnzv committed:
fix(pp): Extra dots in new function calls

jubnzv committed:
feat(pp): Add pp functions

jubnzv committed:
fix(writeRouter): Make old `writeReceiver` work with new API functions

jubnzv committed:
feat(codegen): Port `writeStdlib`

jubnzv commented:
A summary of our recent conversation with @anton-trunov:
* This PR should be merged iteratively. The main blocker is that the old backend is not sufficiently covered by tests. Thus, it is not realistic to provide good differential testing to ensure everything works as expected in the new backend.
* We should use functions from `syntaxConstructors` to map Tact expressions/statements to FunC (`writeFunction.ts` and `writeExpression.ts`). Other FunC functions implicitly introduced by the backend should be parsed as a standard library using the FunC parser. Using AST generator functions there makes the code less readable and maintainable.
* The new backend should be merged and maintained at the same time as the old backend. However, we need to enable it in `tact.config.json`, not with an environment variable as we do now.
* We could use diff in differential tests on the artifacts generated by the compiled code, not on the FunC itself.

Some goals we would like to achieve in the new backend later:
* Some code generation functions might be optimized to reduce code size and gas consumption.
* Refactoring to make it more maintainable.

Thus, I suggest limiting the scope of this backend to the following:
* [x] FunC parser
* [ ] Test suite for differential testing
* [ ] New backend, which is essentially a copy of the old one, implementing the following features:
  + [x] `syntaxConstructors.ts`
  + [ ] An option in the configuration file to choose the backend
  + [ ] Generators of implicit code ported to the new parser:
    - [x] `writeStdlib.ts`
    - [ ] `writeRouter.ts`
  + [ ] Generators that use `syntaxConstructors.ts` to emit source code
    - [ ] Expressions
    - [ ] Statements
    - [ ] Functions
  + [ ] Unit tests for diff testing that cover the implemented logic

After implementing this, I suggest to merge this PR and add more functionality to it iteratively, covering it with tests.

novusnota commented:
> However, we need to enable it in tact.config.json, not with an environment variable as we do now.

How about adding the new experimental option for this. Like, `newFuncBackend`:

```json
{
  "projects": [
    {
      "name": "some_prefix",
      "path": "./contract.tact",
      "output": "./contract_output",
      "options": {
        "experimental": {
          "newFuncBackend": true
        }
      }
    }
  ]
}
```

P.S.: Whoops, a miss-click — didn't mean to close the PR, sorry!
